Learn HTML (modules 12-20) (#9576)
GoogleChrome/web.dev: src/site/content/en/learn/html/audio-video/index.md

authors:
  - estelleweyl
description: Discover how to work with HTML media such as audio and video.
date: 2023-14-02
tags:
  - html
---

As you learned in the [images](/learn/html/images/) module, HTML enables embedding media into a web page. In this section,
we look at audio and video files, including how to embed them, user controls, providing a static image placeholder for your videos,
and best practices, including making audio and video files accessible.

## `<audio>` and `<video>`

The [`<video>`](https://developer.mozilla.org/docs/Web/HTML/Element/video) and [`<audio>`](https://developer.mozilla.org/docs/Web/HTML/Element/audio)
elements can be used to embed media players directly with the `src` attribute or can be used as the container element for a series of [`<source>`](https://developer.mozilla.org/docs/Web/HTML/Element/source) elements,
each providing a `src` file suggestion. While `<video>` can be used to embed an audio file, the `<audio>` element is preferable for embedding
sound files.

> The opening `<video>` and `<audio>` tags can contain several other attributes including global attributes:
> `controls`, `autoplay`, `loop`, `mute`, and `preload`. The `<video>` element also supports the `height`, `width`, and
> `poster` attributes.

Reviewer: ```suggestion
The opening `<video>` and `<audio>` tags can contain several other attributes including 
`controls`, `autoplay`, `loop`, `mute`, `preload`, and the global attributes. The `<video>` element also supports the `height`, `width`, and
`poster` attributes.
```

```html
<video src="videos/machines.webm" poster="images/machine.jpg" controls>
  <p>Watch <a href="https://youtube.com/link">video on Youtube</a>
</video>
```

This `<video>` example has a single source with the `src` attribute linking to the video source. The `poster` attribute
provides an image to display as the video loads. Finally, the `controls` attribute provides user video controls.

> Fallback content is included between the opening and closing tags. If the video (or audio) is not supported or otherwise
> doesn't load, this content is shown. The closing `</video>` tag is required, even if there is no content between the two

Reviewer: ```suggestion
Fallback content is included between the opening and closing tags. If the user agent doesn't support `<video>` (or `<audio>` this content is shown. The closing `</video>` tag is required, even if there is no content between the two
```

from fetching media files it wouldn't be able to decode.

Within the `type` attribute, you can include a [`codecs`](https://developer.mozilla.org/docs/Web/Media/Formats/codecs_parameter) parameter,
which specifies exactly how the resource is encoded. Codecs give you a way of including media optimizations that are not yet supported
in all browsers. The codec is separated from the media type with a semicolon. For example, the codec can be written using
intuitive syntax, such as `<source src="videos/machines.webm" type="video/webm;codecs=vp8,vorbis">` which indicates that the
WebM files contain VP8 video and vorbis audio. Codecs can also be more difficult to decipher, such as `<source src="videos/machines.mp4" type="video/mp4"; codecs=avc1.4d002a">`
which indicates that the MP4 encoding is Advanced Video Coding Main Profile Level 4.2. Explaining this syntax is well beyond
the scope of this lesson. Jake Archibald has a post explaining [how to determine the codec parameter for an AV1 video](https://jakearchibald.com/2022/html-codecs-parameter-for-av1/)
if you want to learn more.

> When displaying a video, by default, the first frame of the video is shown as the still shot, or "poster", when it becomes available.

Reviewer: ```suggestion
When displaying a video, by default, the first frame of the video is shown as the still shot when it becomes available.
```

> This is something that can be controlled. The `poster` attribute enables the source of an image to show while the video is downloading
> and until it is played. Make sure to define the aspect ratio of your video, because when the video loads, a size difference
> between the poster and the video will cause a reflow and repaint.
> 
> In case you were wondering, if the video plays and is subsequently paused, the poster is not re-shown.
> 

Reviewer: ```suggestion
This is something that can be controlled. The `poster` attribute enables the source of an image to show while the video is downloading
and until it is played. If the video plays and is subsequently paused, the poster is not re-shown.

Make sure to define the aspect ratio of your video, because when the video loads, a size difference
between the poster and the video will cause a reflow and repaint.

```

Always include the [boolean](/learn/html/attributes/#boolean-attributes) `controls` attribute. This makes the user controls
visible, providing users with the ability to control audio levels, mute the audio completely, and expand the video to fullscreen.
Omitting `controls` creates a bad user experience, especially if the boolean `autoplay` attribute is included. Note that some
browsers will not heed the `autoplay` attribute directive if the boolean `muted` attribute is omitted because autoplaying a
media file is generally a bad user experience, even when muted and with visible controls.

## Tracks

Between the opening and required closing tags of both audio and video, include one or more [`<track>`](https://developer.mozilla.org/docs/Web/HTML/Element/track)
elements to specify timed text tracks. The following example includes two `<track>` files, providing timed text subtitles in both English and French.

> ```html
> <track label="English" kind="subtitles" srclang="en" src="vtt/subtitles-en.vtt" default />
>    <track label="Francais" kind="subtitles" srclang="fr" src="vtt/subtitles-fr.vtt" />
> ```

Reviewer: ```suggestion
```html
<track label="English" kind="subtitles" srclang="en" src="vtt/subtitles-en.vtt" default />
<track label="Français" kind="subtitles" srclang="fr" lang="fr-fr" src="vtt/subtitles-fr.vtt" />
```
```

The track files, specified in the `src` attribute, should be in [WebVTT format](https://developer.mozilla.org/docs/Web/API/WebVTT_API) (.vtt).
The files should be on the same domain as the HTML document, unless the [`crossorigin`](https://developer.mozilla.org/docs/Web/HTML/Attributes/crossorigin)
attribute is included.

There are five enumerated values for the track `kind` attribute: `subtitles`, `captions`, `descriptions`, `chapters`, and other `metadata`.

> Include `subtitles` along with the `srclang` attribute for dialogue transcription and translations. The value of each label attribute

Reviewer: ```suggestion
Include `subtitles` along with the `srclang` attribute for dialogue transcription and translations. The value of each `label` attribute
```

To improve the performance of always muted videos, [remove the audio track](https://gist.github.com/liangfu/97f877e311210fa0ae18a31fdd92982e) from your media sources.

If your video plays for five seconds or less, accessibility guidelines don't require a pausing mechanism, but anything
with the boolean `loop` attribute will loop forever by default, exceeding this five-second, or any other, time limit. For good
user experience, always include a method of pausing the video. This is most easily done by including `controls`.

## Custom media controls

To display custom video or audio controls, rather than the browser built-in controls, include the `controls` attribute. Then use
JavaScript to add custom media controls and remove the controls attribute. For example, you can add a `<button>` that toggles
the play state of an audio file.

```html
<button id="playPause" aria-controls="idOfAudio"
   data-pause-text="Pause audio"
   data-play-text="Play audio">Pause audio</button>
```

This example includes a button with `dataset` attributes containing the text that will be updated as the visitor toggles

> between play and pause states. The `aria-controls` attribute is included with the id of the element controlled by the button;

Reviewer: ```suggestion
between play and pause states. The `aria-controls` attribute is included with the `id` of the element controlled by the button;
```

in this case, the audio. Each button that controls the audio has an event handler.

> The [`HTMLMediaElement`](https://developer.mozilla.org/docs/Web/API/HTMLMediaElement) provides several properties, methods, and events that are inherited by both the
> [HTMLVideoElement](https://developer.mozilla.org/docs/Web/API/HTMLVideoElement) and
> [HTMLAudioElement](https://developer.mozilla.org/docs/Web/API/HTMLAudioElement), with `HTMLMediaElement` adding a few properties,
> methods, and events of its own. There are several other [Media APIs](https://developer.mozilla.org/docs/Web/Media#apis),
> including a [TextTrack API](https://developer.mozilla.org/docs/Web/API/TextTrack). You can use the [Media Capture and Streams](https://developer.mozilla.org/docs/Web/API/Media_Capture_and_Streams_API)
> and [MediaDevices](https://developer.mozilla.org/docs/Web/API/MediaDevices) APIs to [record audio from a user's microphone](/patterns/media/microphone-record/)
> or [record a user's screen](/patterns/media/screen-record/). The [Web Audio API](https://developer.mozilla.org/docs/Web/API/Web_Audio_API)
> enables [manipulating live and pre-recorded](/patterns/media/audio-effects/) audio and streaming, saving, or sending the audio to the `<audio>` element.
> 

Reviewer: ```suggestion

```
moved to end

```javascript
const pauseButton = document.getElementById('playPause');

pauseButton.addEventListener("click", pauseAndPlay, false);

function pauseAndPlay() {
    console.log(this);
    const media = document.getElementById(this.getAttribute('aria-controls'));

    if (media.paused) {
        media.play();
        this.innerText = this.dataset.pauseText;
    } else {
        media.pause();
        this.innerText = this.dataset.playText;
    }
}
```

By including the `controls` attribute, the user has a way to control the audio (or video) even if JavaScript fails.
Only remove the controls attribute once a replacement button has been instantiated.

```javascript
document.querySelector('[aria-controls]').removeAttribute('controls');
```

Always include external controls when users can't access the controls, such as with background videos that have their controls hidden
behind site content. It is important to understand the basics of [media accessibility requirements](https://www.w3.org/WAI/WCAG21/Understanding/time-based-media) to accommodate users with
different environmental and sensory needs, including the millions of people with hearing loss and visual impairments.

> 

Reviewer: ```suggestion
We've just touched on the [`HTMLMediaElement`](https://developer.mozilla.org/docs/Web/API/HTMLMediaElement) which provides several properties, methods, and events that are inherited by both the
[HTMLVideoElement](https://developer.mozilla.org/docs/Web/API/HTMLVideoElement) and
[HTMLAudioElement](https://developer.mozilla.org/docs/Web/API/HTMLAudioElement), with `HTMLMediaElement` adding a few properties,
methods, and events of its own. There are several other [Media APIs](https://developer.mozilla.org/docs/Web/Media#apis),
including a [TextTrack API](https://developer.mozilla.org/docs/Web/API/TextTrack). You can use the [Media Capture and Streams](https://developer.mozilla.org/docs/Web/API/Media_Capture_and_Streams_API)
and [MediaDevices](https://developer.mozilla.org/docs/Web/API/MediaDevices) APIs to [record audio from a user's microphone](/patterns/media/microphone-record/)
or [record a user's screen](/patterns/media/screen-record/). The [Web Audio API](https://developer.mozilla.org/docs/Web/API/Web_Audio_API)
enables [manipulating live and pre-recorded](/patterns/media/audio-effects/) audio and streaming, saving, or sending the audio to the `<audio>` element.

```

> ## Check your understanding
> 

Reviewer: ```suggestion
```
